File: network/src/main/java/com/mtime/base/signal/SocketManager.java
package com.mtime.base.signal;

import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import androidx.annotation.NonNull;
import android.util.Log;

import com.mtime.base.utils.CollectionUtils;

import org.json.JSONException;
import org.json.JSONObject;

import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;

import io.socket.client.Ack;
import io.socket.client.IO;
import io.socket.client.Socket;
import io.socket.emitter.Emitter;
import okhttp3.Cookie;
import okhttp3.CookieJar;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;

/**
 * Created by yuhengyi on 2017/7/19.
 * <p>
 * 匹配房间 加入房间 离开房间 取消匹配等 长连接
 * 看代码前 要先阅读 信令wiki文档http://gitlab.mtime-dev.com/short-video/svs-room/wikis/home
 */

public class SocketManager {

    public static final String WEAK_UP_SOCKET = "mtimeinterweakupsocket";

    private static final int RECONNECT_DELAY = 1500;
    // 发送消息 超时
    private static final long SOCKET_ACK_TIMEOUT = 5000;
    private static final int SOCKET_TIMEOUT = -2;

    private Socket sSocket;
    private static OkHttpClient okHttpClient;

    /**
     * 是否需要主动重连
     */
    private boolean mNeedReconnect = false;

    private static final ConcurrentHashMap<String, List<Cookie>> sCookies = new ConcurrentHashMap<>();

    private final Handler mReconnectHandler;

    private boolean mIsUserDisconnect;

    private final JSONObject mInnerError = new JSONObject();

    SocketManager(String host) {
        mReconnectHandler = new ReconnectHandler(LooperThread.get().getLooper());
        IO.Options options = new IO.Options();
        if (okHttpClient == null) {
            OkHttpClient.Builder builder = new OkHttpClient.Builder();
            builder.cookieJar(new CookieJar() {

                private String url(HttpUrl url) {
                    return url.scheme() + "://" + url.host() + ":" + url.port() + url.encodedPath();
                }

                @Override
                public void saveFromResponse(@NonNull HttpUrl url, @NonNull List<Cookie> cookies) {
                    if (CollectionUtils.size(cookies) > 0) {
                        sCookies.put(url(url), cookies);
                    }
                }

                @Override
                public List<Cookie> loadForRequest(@NonNull HttpUrl url) {
                    List<Cookie> cookies = sCookies.get(url(url));
                    return cookies != null ? cookies : new ArrayList<Cookie>();
                }
            });
            okHttpClient = builder.build();
        }
        options.webSocketFactory = okHttpClient;
        options.callFactory = okHttpClient;
        options.transports = new String[]{
                "websocket"
        };
        options.reconnection = false;
        try {
            sSocket = IO.socket(host, options);
        } catch (URISyntaxException e) {
            e.printStackTrace();
        }
    }

    public Looper getLooper() {
        return LooperThread.get().getLooper();
    }

    void onConnected(final ConnectedListener listener) {
        sSocket.on(Socket.EVENT_CONNECT, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                mReconnectHandler.removeCallbacksAndMessages(null);
                listener.onConnected(args);
            }
        });
    }

    void onDisconnect(final DisconnectListener listener) {
        sSocket.on(Socket.EVENT_DISCONNECT, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                if (!mIsUserDisconnect) {
                    mReconnectHandler.sendEmptyMessageDelayed(1, RECONNECT_DELAY);
                }
                listener.onDisconnect(args);
            }
        });
    }

    public void on(String event, EmitterListener emitterListener) {
        sSocket.on(event, emitterListener);
    }

    public void send(String event, Map<String, Object> params) {
        send(event, params, null);
    }

    public void send(String event, Map<String, Object> params, final EventListener listener) {
        if (mNeedReconnect) return;
        if (!sSocket.connected()) {
            connect();
            return;
        }
        if (event.equals(WEAK_UP_SOCKET)) return;
        JSONObject obj = buildParams(params);
        Object[] objects = new Object[]{obj};
        emit(event, objects, listener);
    }

    @SuppressWarnings("unchecked")
    private JSONObject buildParams(Map<String, Object> params) {
        JSONObject obj = new JSONObject();
        if (params == null) return obj;
        Iterator<String> iterator = params.keySet().iterator();
        try {
            while (iterator.hasNext()) {
                String key = iterator.next();
                Object value = params.get(key);
                if (value instanceof Map) {
                    JSONObject jObj = buildParams((Map<String, Object>) value);
                    obj.put(key, jObj);
                } else {
                    obj.put(key, value);
                }
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return obj;
    }


    public void send(String event, JSONObject jsonObject) {
        send(event, jsonObject, null);
    }

    public void send(String event, JSONObject jsonObject, final EventListener listener) {
        if (mNeedReconnect) return;
        if (!sSocket.connected()) {
            connect();
            return;
        }
        if (event.equals(WEAK_UP_SOCKET)) return;
        Object[] objects = new Object[]{jsonObject};
        emit(event, objects, listener);
    }

    private void emit(String event, Object[] objects, EventListener listener) {
        SocketTimeoutAck ack = null;
        if (listener != null) {
            ack = new SocketTimeoutAck(listener);
            mReconnectHandler.postDelayed(ack, SOCKET_ACK_TIMEOUT);
        }
        sSocket.emit(event, objects, ack);
    }

    /**
     * 连接socket
     */
    void connect() {
        if (sSocket == null) {
            return;
        }
        sSocket.connect();
        mReconnectHandler.sendEmptyMessageDelayed(1, RECONNECT_DELAY);
        mIsUserDisconnect = false;
        mNeedReconnect = false;
    }

    public void cancelAttemptReconnet() {
        mIsUserDisconnect = true;
    }

    /**
     * 主动重连
     */
    void reconnect() {
        if (mNeedReconnect) {
            connect();
            mNeedReconnect = false;
        }
        mIsUserDisconnect = false;
        mReconnectHandler.sendEmptyMessageDelayed(1, RECONNECT_DELAY);
    }

    /**
     * 断开
     */
    void disconnect() {
        if (null != sSocket) {
            sSocket.disconnect();
        }
        mIsUserDisconnect = true;
        mReconnectHandler.removeCallbacksAndMessages(null);
        mNeedReconnect = true;
    }

    /**
     * 释放资源
     */
    void release() {
        if (sSocket != null) {
            sSocket.disconnect();
        }
        mReconnectHandler.removeCallbacksAndMessages(null);
        mIsUserDisconnect = true;
        sCookies.clear();

        mNeedReconnect = false;
    }

    private class ReconnectHandler extends Handler {

        private ReconnectHandler(Looper looper) {
            super(looper);
        }

        @Override
        public void handleMessage(Message msg) {
            if (sSocket != null && !sSocket.connected()) {
                sSocket.connect();
                Log.i("socket", "reconnect");
                sendEmptyMessageDelayed(1, 1500);
            }
        }
    }

    private class SocketTimeoutAck implements Ack, Runnable {

        EventListener mListener;
        // 原子操作，防止多线程异常
        private final AtomicBoolean handling = new AtomicBoolean(false);

        SocketTimeoutAck(EventListener listener) {
            mListener = listener;
        }

        @Override
        public void call(Object... args) {
            if (!handling.compareAndSet(false, true)) return;
            mReconnectHandler.removeCallbacks(this);
            JSONObject result = (JSONObject) args[0];
            String msg = result.has("showMsg") ?
                    result.optString("showMsg", "")
                    : result.optString("msg", "");
            if (result.has("code") && result.optInt("code") == 1) {
                if (mListener != null)
                    mListener.onSuccess(msg, result);
            } else {
                if (mListener != null) {
                    if (mListener instanceof DataEventListener) {
                        DataEventListener del = (DataEventListener) mListener;
                        del.onFailure(msg, result);
                    } else {
                        mListener.onFailure(msg, result.optInt("code"));
                    }
                }
            }
            mListener = null;
        }

        @Override
        public void run() {
            if (!handling.compareAndSet(false, true)) return;
            if (mListener != null) {
                String msg = "socket send timeout: -2";
                if (mListener instanceof DataEventListener) {
                    fillInnerMsgCode(msg, SOCKET_TIMEOUT);
                    DataEventListener del = (DataEventListener) mListener;
                    del.onFailure(msg, mInnerError);
                } else {
                    mListener.onFailure(msg, SOCKET_TIMEOUT);
                }
            }
            mListener = null;
        }
    }

    private void fillInnerMsgCode(String msg, int code) {
        try {
            mInnerError.put("code", code);
            mInnerError.put("showMsg", msg);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public interface EventListener {
        void onSuccess(String msg, JSONObject data);

        void onFailure(String msg, int code);
    }

    public static abstract class DataEventListener implements EventListener {
        public void onFailure(String msg, JSONObject data) {
            onFailure(msg, data.optInt("code"));
        }

        @Override
        public void onFailure(String msg, int code) {
        }
    }

    public static abstract class EmitterListener implements Emitter.Listener {

        @Override
        public final void call(Object... args) {
            JSONObject result = (JSONObject) args[0];
            try {
                onRecive(result);
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }

        public abstract void onRecive(JSONObject data) throws JSONException;
    }

    interface ConnectedListener {
        void onConnected(Object... args);
    }

    interface DisconnectListener {
        void onDisconnect(Object... args);
    }
}

File: api/src/main/java/com/kotlin/android/api/socket/SocketManager.kt
package com.kotlin.android.api.socket

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.annotation.NonNull
import com.kotlin.android.core.CoreApp
import com.kotlin.android.retrofit.ssl.SSL
import com.kotlin.android.retrofit.ssl.UnSafeHostnameVerifier
import io.socket.client.IO
import io.socket.client.Manager
import io.socket.client.Socket
import io.socket.emitter.Emitter
import io.socket.engineio.client.transports.WebSocket
import okhttp3.*
import java.net.URISyntaxException
import java.util.*
import java.util.concurrent.ConcurrentHashMap

/**
 * create by lushan on 2021/2/22
 * description: socket基类
 */
open class SocketManager {

    private var okHttpClient: OkHttpClient? = null
    private var sCookies = ConcurrentHashMap<String, List<Cookie>>()
    var socket: Socket? = null
    private val SEND_MESSAGE_EVENT = "send_message"//发送消息事件
    private val NEW_MESSENGE_EVENT = "new_message"//接收新消息事件

    protected fun init(url: String) {
        if (okHttpClient == null) {
            val builder = OkHttpClient.Builder()
            builder.cookieJar(object : CookieJar {
                private fun url(url: HttpUrl): String {
                    return url.scheme + "://" + url.host + ":" + url.port + url.encodedPath
                }

                override fun saveFromResponse(@NonNull url: HttpUrl, @NonNull cookies: List<Cookie>) {
                    if (cookies.size > 0) {
                        sCookies[url(url).orEmpty()] = cookies
                    }
                }

                override fun loadForRequest(@NonNull url: HttpUrl): List<Cookie> {
                    val cookies: List<Cookie>? = sCookies.get(url(url).orEmpty())
                    return cookies ?: ArrayList()
                }
            })
//            设置ssl
            builder.hostnameVerifier(UnSafeHostnameVerifier())
            SSL.instance.ignoreSSL?.apply {
                builder.sslSocketFactory(sslSocketFactory, trustManager)
            }
            okHttpClient = builder.build()
        }
        val options = IO.Options().apply {
            webSocketFactory = okHttpClient
            callFactory = okHttpClient
            transports = arrayOf(
                    WebSocket.NAME
            )
            reconnection = true
        }

//        创建socket
        try {
            socket = IO.socket(url, options)
        } catch (e: URISyntaxException) {
            e.printStackTrace()
        }
    }

    fun connect() {
        socket?.connect()
    }

    open fun onConnected(vararg args: Any?) {

    }

    open fun onDisconnected() {

    }

    /**
     * 长链接链接状态监听
     */
    fun setSocketListener(listener: ISocketListener) {
//        链接成功
        socket?.on(Socket.EVENT_CONNECT) {
            onConnected(it)
            CoreApp.instance.applicationContext.runOnUiThread {
                Log.e("SocketManager", "链接成功")
                it.forEach { any ->
                    Log.e("SocketManager", "链接成功" + any.toString())
                }

                listener.onConnected()
            }

        }
//        链接错误
        socket?.on(Socket.EVENT_CONNECT_ERROR) {
            CoreApp.instance.applicationContext.runOnUiThread {
                it.forEach { any ->
                    Log.e("SocketManager", "链接错误", any as Throwable)
                }
                listener.onConnectError()
            }

        }
//       链接断开
        socket?.on(Socket.EVENT_DISCONNECT) {
            onDisconnected()
            CoreApp.instance.applicationContext.runOnUiThread {
                it.forEach { any ->
                    Log.e("SocketManager", "链接断开" + any.toString())
                }
                listener.onDisconnect()
            }

        }

        socket?.on(Manager.EVENT_OPEN) {
            CoreApp.instance.applicationContext.runOnUiThread {
                it.forEach { any ->
                    Log.e("SocketManager", "链接成功Open" + any.toString())
                }
                listener.onConnected()
            }

        }
    }

    /**
     * 接收消息事件监听
     * @param listener 接收消息监听事件
     * @param event 接收新消息注册事件名称
     */
    fun setEventListener(event: String = NEW_MESSENGE_EVENT, listener: Emitter.Listener) {
        socket?.on(event, listener)
    }

    /**
     * 发送消息
     * @param content 发送消息内容
     * @param event 发送消息注册事件名称
     */
    open fun sendEvent(event: String = SEND_MESSAGE_EVENT, content: String) {
        socket?.emit(event, content)
    }

    fun disconnect() {
        socket?.disconnect()
        socket?.off()
    }

    fun isConnecting():Boolean = socket?.connected()?:true


    interface ISocketListener {
        fun onConnected()
        fun onConnectError()
        fun onDisconnect()
    }

}

fun Context.runOnUiThread(f: Context.() -> Unit) {
    if (Looper.getMainLooper() === Looper.myLooper()) f() else ContextHelper.handler.post { f() }
}

private object ContextHelper {
    val handler = Handler(Looper.getMainLooper())
}